feat(Instrumentation/Messenger): implement worker and middleware instrumentation

jvocampings opened the pull request:
WIP.

Related to issue https://github.com/FriendsOfOpenTelemetry/opentelemetry-bundle/issues/172

Encapsulate consumption of message in a Span. Based on Messenger events.

Missing instrumentation when a message is dispatched. Not sure how to implement it.

edit: I assume we could reuse TraceableMessengerTransport to decorate the method `send`.

jvocampings committed:
chore(messenger): add trace parent in dispatched messages (AMQP)

jvocampings committed:
chore(messenger): add trace parent in dispatched messages (AMQP)

jvocampings committed:
chore(messenger): use Messenger events to start/end span for instrumentation

jvocampings committed:
chore(messenger): use Messenger events to start/end span for instrumentation

gaelreyrol commented on src/Instrumentation/Symfony/Messenger/EventSubscriber/EndSpanEventSubscriber.php:
I think this subscriber can be merged with the `StartSpanEventSubscriber`, and moved at the root of the Messenger folder.

jvocampings replied:
merged in `InstrumentationEventSubscriber`

gaelreyrol commented on src/Instrumentation/Symfony/Messenger/EventSubscriber/EndSpanEventSubscriber.php:
```diff
@@ -0,0 +1,80 @@
+<?php
+
+declare(strict_types=1);
```
There is no need for strict types.

jvocampings replied:
you don't want it to be sure the bundle is compatible even with apps running without strict_types=1 ?

gaelreyrol replied:
I do not want to infect any application using this bundle with strict types, I prefer to rely on PHPStan to make things right rather than potentially breaking application side code.

You check this issue from Symfony to get severals pov on the subject: https://github.com/symfony/symfony/issues/28423

jvocampings replied:
removed everywhere in the directory `src`

gaelreyrol commented on src/Instrumentation/Symfony/Console/TraceableConsoleEventSubscriber.php:
```diff
@@ -30,6 +30,10 @@ final class TraceableConsoleEventSubscriber implements EventSubscriberInterface,
      */
     private array $excludeCommands = [];
 
+    private array $notSupportedCommands = [
```
Please add phpDoc to type the array.

jvocampings replied:
done

gaelreyrol commented on src/Instrumentation/Symfony/Messenger/EventSubscriber/EndSpanEventSubscriber.php:
```diff
@@ -0,0 +1,80 @@
+    public static function getSubscribedEvents()
+    {
+        return [
+            WorkerRunningEvent::class => ['endSpan'],
+            WorkerStoppedEvent::class => ['endSpan'],
```
We should not end a span on running/stopped worker events but on WorkerMessageReceived and WorkerMessageHandled events.

jvocampings replied:
ended on WorkerMessageHandledEvent. But I don't understand why it should also be ended on WorkerMessageReceived

gaelreyrol replied:
I am sorry I mean, start a span on WorkerMessageReceived event and end span on WorkerMessageHandled event.

gaelreyrol commented on src/Instrumentation/Symfony/Messenger/EventSubscriber/EndSpanEventSubscriber.php:
```diff
@@ -0,0 +1,80 @@
+        $span->end();
+    }
+
+    public function hydrateSpanWithError(WorkerMessageFailedEvent $event): void
```
On failure the span should also be ended.

jvocampings replied:
done

gaelreyrol commented on src/Instrumentation/Symfony/Messenger/TraceableMessengerMiddleware.php:
```diff
@@ -8,6 +8,9 @@
+/**
+ * @deprecated: span should be created from event subscriber to avoid issue with orphan span
+ */
 class TraceableMessengerMiddleware implements MiddlewareInterface
```
I think there is misunderstanding here, we still need the middleware but maybe there is some adjustment to do.
You have to thing about the fact that a message can handled synchronously and it won't trigger worker events.

According to Symfony's documentation:
> The middleware are executed when the message is dispatched but also again when a message is received via the worker (for messages that were sent to a transport to be handled asynchronously). Keep this in mind if you create your own middleware.

We need to find a way to either trace from the middleware when a synchronous message is dispatched or attach the trace context with a stamp when the message dispatched to the transport and trace it when the message is received and handled.

Here is the rolling of events:

1. a message is dispatched
2. the message goes through middlewares -> span start if sync
3. the message is handled if synchronous -> span end if sync
4. the message is sent to transport if asynchronous -> attach tracing context without span creation
5. the message is received by the worker -> span start
6. the message goes through middlewares -> no span creation
7. the message is successfully handled by the worker -> span end
8. the messagge has failed to be handled by the worker -> span end

By reviewing this, I think we could allow span creation within the middleware in an asynchronous context too as it will detail the every phases of the message, the dispatch phase and the handling phase as with synchronous message the dispatch/handling phase happen at the same time.

Still, we need to do some experiments, in each case to track down spans and make sure they are relevant, well linked and ended correctly.

gaelreyrol commented:
> edit: I assume we could reuse TraceableMessengerTransport to decorate the method send.
You mean attach the trace context to the message?

Thanks again for your preliminary work on this, it's not an easy part and it will take some time to get the right implementation!

jvocampings committed:
Merge remote-tracking branch 'origin/messenger-propagation' into messenger-instrumentation

jvocampings commented:
> > edit: I assume we could reuse TraceableMessengerTransport to decorate the method send.
> > You mean attach the trace context to the message?

Ah I was talking about Instrumentation and not Propagation. 
When I dispatch a message which is consumed asynchronously, I'd like to create a span to have a trace of it and to see how many time it took to send messages to my broker (rabbitmq, sqs etc)

jvocampings committed:
remove strict type, merge start/end span into 1 subscriber, close span on error and on message handled

jvocampings committed:
remove strict type, merge start/end span into 1 subscriber, close span on error and on message handled

jvocampings commented:
@gaelreyrol I made some modifications related to what you asked. I'll be on holiday for 2 weeks. After that there are big changes at my work then I don't know if i'll be able to reactive. I'll do my best.

jvocampings committed:
propagation not related to AMQP + read incoming trace (async context)

jvocampings committed:
propagation not related to AMQP + read incoming trace (async context)

gaelreyrol commented:
> Ah I was talking about Instrumentation and not Propagation. When I dispatch a message which is consumed asynchronously, I'd like to create a span to have a trace of it and to see how many time it took to send messages to my broker (rabbitmq, sqs etc)

It can't be done from the event and the `TraceableMessengerTransport` is already doing that, you just need to wrap the transport DNS with the `trace()` decorator. Check out this file: https://github.com/FriendsOfOpenTelemetry/opentelemetry-bundle/blob/main/tests/Functional/Application/config/packages/framework.yaml#L14

> I  made some modifications related to what you asked. I'll be on holiday for 2 weeks. After that there are big changes at my work then I don't know if I'll be able to reactive. I'll do my best.

Thank you very much, don't worry at all, have a good holiday! I might work on your PR directly if I have enough time.

gaelreyrol commented on src/Instrumentation/Symfony/Messenger/InstrumentationEventSubscriber.php:
```diff
@@ -0,0 +1,128 @@
+ * Be aware if you start a span before this subscriber, it could leads to orphan span issue.
+ * Be sure your span is properly ended.
+ */
+class InstrumentationEventSubscriber implements EventSubscriberInterface
```
Since this subscriber is about worker events, I think we should name it `WorkerMessageEventSubscriber`.

gaelreyrol commented on src/Instrumentation/Symfony/Messenger/InstrumentationEventSubscriber.php:
```diff
@@ -0,0 +1,128 @@
+
+    public function startSpan(WorkerMessageReceivedEvent $event): void
+    {
+        if (InstrumentationTypeEnum::Auto !== $this->instrumentationType) {
```
I'd like to adapt this to handle manual instrumentation, for example if the message has the `#[Traceable]` attribute, it should be instrumented.

jvocampings committed:
rename event subscriber

jvocampings committed:
rename event subscriber

gaelreyrol committed:
fix(messenger): clean up worker subscriber and add functional tests

Fix several issues in the WorkerMessageEventSubscriber introduced by PR #173:
- Replace SDK Span import with API Span to respect API/SDK separation
- Implement InstrumentationTypeInterface for consistency with other subscribers
- Add event priorities (10000/-10000) to wrap all other processing
- Add messaging semantic convention attributes (operation.type, destination.name)
- Include message class name in span name for better trace readability
- Remove stale imports and duplicate propagation middleware service definition
- Clean up propagation middleware when messenger tracing is disabled
- Add PHPStan baseline entries for untyped $carrier interface params
- Add functional tests for worker message handled, failed, and attribute mode
- Reorganize messenger tests into Messenger/ subdirectory
- Disable retry on test transport to isolate worker span assertions

gaelreyrol committed:
test(messenger): add tests for transport tracing and propagation middleware

Cover TraceableMessengerTransport (get/ack/reject spans + TransportException
error recording) and AddStampForPropagationMiddleware (stamp skip, no-scope
passthrough, active-scope injection).

gaelreyrol committed:
Merge branch 'messenger-instrumentation' into jvocampings-messenger-instrumentation

gaelreyrol committed:
fix(messenger): use recordException() instead of manual attributes

Align WorkerMessageEventSubscriber.endSpanOnError() with every other
subscriber in the codebase (HttpKernel, Console, Doctrine, Mailer,
TransportTracer) by using $span->recordException() which creates a
proper OTel exception event with exception.type, exception.message,
and exception.stacktrace per the OTel semantic conventions.

Update tests to assert on span events rather than span attributes.

gaelreyrol committed:
fix(messenger): unwrap HandlerFailedException in error spans

When Symfony Messenger handlers fail, the exception is wrapped in a
HandlerFailedException (or other WrappedExceptionsInterface implementors
like DelayedMessageHandlingException). Without unwrapping, the span
records the generic wrapper message instead of the actual root causes.

Now each nested exception is recorded as a separate span event via
recordException(), matching how Sentry handles these wrapped exceptions.

gaelreyrol committed:
fix(console): make messenger:consume exclusion configurable

Replace the hardcoded $notSupportedCommands array and isNotSupported()
method with a default value on the existing exclude_commands config node.

messenger:consume is now the default entry in console.tracing.exclude_commands,
following the same pattern as Sentry's configurable excluded_commands. Users
can override the list or clear it entirely via configuration.

gaelreyrol committed:
test(messenger): add unit test for TraceStampPropagator

Cover all branches of set(), get(), and keys() methods:
- Valid envelope carrier for set/get/keys
- Invalid carrier throws InvalidArgumentException
- Non-traceparent key is ignored (set) or returns null (get)
- Missing TraceStamp returns null from get()

Closes the 34.78% patch coverage gap reported by Codecov.

gaelreyrol committed:
fix(messenger): propagate tracestate alongside traceparent

The W3C Trace Context spec defines two headers: traceparent and
tracestate. Previously only traceparent was propagated, silently
dropping vendor-specific trace context (Datadog, AWS X-Ray, etc.).

Add optional traceState property to TraceStamp and update
TraceStampPropagator to handle both keys in set(), get(), and keys().
When tracestate arrives after traceparent, the existing stamp is
replaced with one carrying both values.

gaelreyrol committed:
fix(messenger): namespace bus.name attribute as symfony.messenger.bus.name

bus.name is not a standard OTel semantic convention. Rename to
symfony.messenger.bus.name to follow the existing codebase pattern
(symfony.console.*, symfony.kernel.*).

Only changes the worker subscriber attribute; the pre-existing
bus.name in TraceableMessengerStack is left unchanged (out of scope).

gaelreyrol committed:
fix(messenger): add scope leak safety in worker subscriber

If startSpan fires but neither endSpanWithSuccess nor endSpanOnError
fires (e.g. worker killed, unhandled error in another subscriber),
the OTel context scope leaks into subsequent messages.

Now startSpan checks for a lingering scope at the beginning of each
message and cleans it up: detaches the scope, marks the orphaned span
as ERROR, and ends it. This prevents context pollution across messages.

gaelreyrol committed:
feat(messenger): add retry awareness, attribute instrumentation, and namespace span attributes

- Namespace middleware span attributes (bus.name, event.category, event.current)
  under symfony.messenger.* prefix for consistency with worker subscriber
- Add symfony.messenger.will_retry and symfony.messenger.retry_count attributes
  to failure spans so operators can distinguish retriable from terminal failures
- Support #[Traceable] attribute on message classes for attribute-based
  instrumentation mode, with custom tracer selection via TracerLocatorPass

gaelreyrol committed:
refactor(messenger): improve type safety, observability, and test coverage

Mark all new types as final for consistency with existing codebase.
Register TraceStampPropagator as a shared DI service instead of inline
construction. Default instrumentationType to Auto, add safe tracer
locator fallback, nullable logger consistency, debug logging on skipped
messages and null scopes, and remove dead commented-out code. Add
end-to-end propagation test, WorkerMessageEventSubscriber unit tests,
and PropagatorFactory test.

gaelreyrol committed:
fix(http-client): use lowercase header key for Content-Length lookup

Symfony's getHeaders() normalizes header names to lowercase, so the
Content-Length check was unreachable. Access the first array element
since headers are returned as arrays.

gaelreyrol committed:
fix(messenger): use explicit scope tracking and record errors on middleware spans

TraceableMessengerStack now tracks its own scope, span, and parent context
instead of probing global context storage — fixing scope leaks where earlier
middleware spans were never ended. TraceableMessengerMiddleware catches
exceptions and records them with STATUS_ERROR on the active span, aligning
with every other instrumentation in the bundle.

gaelreyrol committed:
fix(messenger): remove erroneous kernel.reset tag from transport factory

TraceableMessengerTransportFactory was tagged `kernel.reset` with
method `reset`, but the class has no such method. Symfony's
ServicesResetter invokes the declared method on every tagged service
between worker iterations, so `messenger:consume` threw as soon as a
retry (or any post-handler reset) ran:

    Attempted to call an undefined method named "reset" of class
    "...\TraceableMessengerTransportFactory".

The tag was introduced with the original messenger tracing (6f3c032)
and never had a corresponding method — likely speculative, possibly
confused with messenger transports that sometimes implement
ResetInterface.

Removal is preferred over adding an empty reset(): the factory's
dependencies (inner TransportFactory, TracerInterface, LoggerInterface)
are immutable DI services, createTransport() returns a fresh
TraceableMessengerTransport without retaining a reference, and
supports() is pure — there is no per-message state to clear. An empty
reset() would silence the error but imply a ResetInterface-style
contract the class does not fulfill and mislead anyone later adding
mutable state.

gaelreyrol committed:
Merge branch 'main' into jvocampings-messenger-instrumentation

gaelreyrol commented:
Hi @jvocampings — quick update on where the PR stands. cc @kochen (ref #190) — this should cover your use case too; would love your feedback.

**Review feedback addressed**
- Merged `StartSpan`/`EndSpan` subscribers into a single `WorkerMessageEventSubscriber` at the root of the Messenger folder
- Removed `strict_types` throughout `src/`
- Span lifecycle now hooks `WorkerMessageReceivedEvent` (start) and `WorkerMessageHandledEvent`/`WorkerMessageFailedEvent` (end)
- Error path ends the span and records the exception via `recordException()`; `HandlerFailedException` is unwrapped so the root cause surfaces in traces

**Middleware & propagation**
- Kept the middleware per the sync/async flow we discussed — it creates spans in both paths so dispatch and handling phases are visible
- Trace context is propagated through a dedicated stamp carrying both `traceparent` and `tracestate` (no longer AMQP-specific)
- Explicit scope tracking with leak safety; errors are recorded on middleware spans too

**Attributes, config, tests**
- Span attributes namespaced as `symfony.messenger.*`
- `messenger:consume` exclusion is now configurable (Console instrumentation)
- Added functional tests for the worker subscriber and unit tests for `TraceStampPropagator`, transport tracing, and the propagation middleware
- Drive-by: fixed an erroneous `kernel.reset` tag on the transport factory and a lowercase `Content-Length` lookup in HttpClient

@jvocampings — could you give the branch a spin on your setup to confirm nothing regressed vs. your original intent? And @kochen, if you're able to test against your PDF-generator flow, that would be a great real-world validation.

gaelreyrol committed:
Merge branch 'main' into jvocampings-messenger-instrumentation

gaelreyrol commented:
@jvocampings @kochen I am merging this pull request along with the issues that should be fixed with this merge. Don't hesitate to reopen them!

gaelreyrol commented:
Closes #172 #190
